### ISSUE TITLE
Not exactly an issue, but a question, will you publish a patch at developer.blender.org?

### ISSUE THREAD

**Issue by juangea** [JiahuiCai/Blender_Interactive_Motion_Path#1]
This looks amazing, would be awesome to have it commited in master, will you publish a patch in developer.blender.org? :D

**Comment by JiahuiCai** [comment]
Thanks, but unfortunately I don't plan to do so since this tool has limitations so it is not production ready. However, I am sure a subset of the functionalities of this tool can make it into the main build.

My original intention is to see if this workflow can replace the whole process of rigging so animators can directly work with the vanilla FK rig instead of setting up something like rigify. But there are use cases that FK rig just won't cut it, for example, moving the hip while pinning the feet is very tedious to do using this workflow. So the fact that this tool gives FK rig IK behavior is not that useful in the production.

However, editing the motion path of the IK controllers that are only keyed on location could potentially augment existing blender workflow. Animators still need to build a rig first before animating, but IK to FK switch may no longer be required. for example, if you key two poses using the IK controller, the interpolation will be a straight line. If you ever want to get a curve path, the current workflow is to switch from IK to FK mode on the rig and then key on the Fk controller. This tool can simplify this process by allowing you to create a curved path between two IK poses.

Giving IK rig FK-like interpolation is a trivial functionality that is easy to do. I could help make it happen.

**Comment by JiahuiCai** [comment]
Giving IK rig FK-like interpolation can also be accomplished by an existing addon
https://www.youtube.com/watch?v=QNTvEiD5bzE
https://blendermarket.com/products/motiontrail3d

**Comment by juangea** [comment]
Thanks! 

I just noticed your words in Blender Chat.

The most important part I was referring to are the motion paths, they are amazing! the only thing I miss for them are some bezier handlers to be able to modify the path without the need of going into the graph editor.

The part I didn't liked too much is the part of pinnig FK, while theoretically it could work, it's a bit tedious for an animator to have to be pinning things all the time, but that's my personal opinion.

**Comment by juangea** [comment]
The motion trail addon is what I mean, but I'm afraid that being an addon it could be super slow in a complex rig :)

**Comment by JiahuiCai** [comment]
Pinning FK is definitely a tedious workflow I can assure you now that I have tried with my tool, there are ways around it such as when you want to move the hip, you add the feet to an automatic pinning set that is auto-pinned, sounds kind of like "Descriptive rigging" and then you can have multiple pinning set that can be activated/deactivated on the fly which then gives you the “ephemeral” rigging kind of system. https://www.youtube.com/watch?v=J48b0GKI4RM
I am not sure how well it is gonna work for more complex rigs like creature rigs tho.

Basically what I want to achieve is that once the user created the FK rig and skinning it, the user should be able to start animating right away. Rigging should not be part of the process.

And from that point on, for every user's intended action, there is a tool /operator for that instead of having to build a rig that supports every action from the very beginning. whether it be IK/FK posing, interpolate between two poses in a certain way, apply an additive pose to a range of frames, run a simulation on a range of frames for selected bones. 

But this is for the future, for now, I think for a regular workflow using a production rig, the best option now is the motion trail 3d addon I think.

**Comment by juangea** [comment]
I understand your target, and it could be useful for animation tests and such things, but for an animation production, the level of control required by the animators involve much more than the body bones, with a lot of tiny bones that control a lot of small things, or maintain volume in place, or deform volume in a special case, so it could be useful, but a rig is usually needed for such things :)

Can't the "motion trail" part of this development be stripped out? at least the real time drawing I mean.

**Comment by JiahuiCai** [comment]
Definitely, I think I miss use the word rigging here, I was talking about things like IK, foot rolls, FK/IK switch/snap, spline IK, all the good stuff for big motions, primary bones. Secondary bones and shape keys are definitely required, but I think there are very few dependencies between secondary bones and primary bones, and they are usually separated by bone layers. But you are right, this part of rigging is 100% needed.

Yes, the part that draws the path and the part that makes it editable for bones that are keyed on locations can be stripped out. I can look into it in the future. But the existing addon already does this really well. Do you have performance issue with MotionTrail3D addon? The dev for that addon is actively looking to fix performance issues. I am also curious too.

Integrating with blender officially has two challenges. My work is independent of the existing motion path code, everything has to be rewritten if the solution is to change the existing motion path code to behave like my tool, and as far as I can tell the blender motion path code is designed to be non-interactive from the beginning so a lot of change is required. The second challenge is that if I just push for my tool to be integrated as it is now, then Blender will have two motion trail system, which is definitely not ideal

But for the drawing part, I thought Blender's motion path already supports real-time updates? At the end of each transform the motion path is updated.

**Comment by Lauloque** [comment]
I personally don't like Blender's current motion path tool. It's hidden in a corner I never visit when I'm animating, it requires constant attention to manage the frame range, baking, updating, there's no color coding to show important data like spacing, no relative frame range, no fade, ... No thanks.

My favorite motion trail of the moment is [AnimBot's motion trail](https://www.animbot.ca/tooltips/#motion_trail). It's right under the hand at all times, it's fast, flexible, and once you set what you need, you don't have to babysit it. It just lets you enjoy it.

Besides, is there a way to switch your motion trail gizmo between heads and tails?

**Comment by JiahuiCai** [comment]
> I personally don't like Blender's current motion path tool. It's hidden in a corner I never visit when I'm animating, it requires constant attention to manage the frame range, baking, updating, there's no color coding to show important data like spacing, no relative frame range, no fade, ... No thanks.
> 
> My favorite motion trail of the moment is [AnimBot's motion trail](https://www.animbot.ca/tooltips/#motion_trail). It's right under the hand at all times, it's fast, flexible, and once you set what you need, you don't have to babysit it. It just lets you enjoy it.
> 
> Besides, is there a way to switch your motion trail gizmo between heads and tails?

yes there is a way. select bone, go to bone properties, right below ik settings there are the settings for motion curve. The first two checkbox  allow u to toggle head tail curve display. The last gif in the main page also shows this

**Comment by Lauloque** [comment]
> yes there is a way. select bone, go to bone properties, right below ik settings there are the settings for motion curve. The first two checkbox allow u to toggle head tail curve display. The last gif in the main page also shows this

Indeed, I missed that one thanks :o
